planning/mission_planner/src/mission_planner/route_selector.cpp

```diff
@@ -0,0 +1,218 @@
+}
+
+void RouteSelector::on_clear_route_main(
+  ClearRoute::Request::SharedPtr req, ClearRoute::Response::SharedPtr res)
+{
+  // Save the request to resume from MRM.
+  main_request_ = std::monostate{};
+
+  // During MRM, only change the state.
+  if (mrm_operating_) {
+    main_.change_state(now(), RouteState::UNSET);
+    res->status.success = true;
+    return;
+  }
+
+  // Forward the request if not in MRM.
+  res->status = service_utils::sync_call(cli_clear_route_, req);
+}
+
+void RouteSelector::on_set_waypoint_route_main(
+  SetWaypointRoute::Request::SharedPtr req, SetWaypointRoute::Response::SharedPtr res)
+{
+  // Save the request to resume from MRM.
+  req->uuid = uuid::generate_if_empty(req->uuid);
+  main_request_ = req;
+
+  // During MRM, only change the state.
+  if (mrm_operating_) {
+    main_.change_state(now(), RouteState::SET);
+    res->status.success = true;
+    return;
+  }
+
+  // Forward the request if not in MRM.
+  res->status = service_utils::sync_call(cli_set_waypoint_route_, req);
+}
+
+void RouteSelector::on_set_lanelet_route_main(
+  SetLaneletRoute::Request::SharedPtr req, SetLaneletRoute::Response::SharedPtr res)
+{
+  // Save the request to resume from MRM.
+  req->uuid = uuid::generate_if_empty(req->uuid);
+  main_request_ = req;
+
+  // During MRM, only change the state.
+  if (mrm_operating_) {
+    main_.change_state(now(), RouteState::SET);
+    res->status.success = true;
+    return;
+  }
+
+  // Forward the request if not in MRM.
+  res->status = service_utils::sync_call(cli_set_lanelet_route_, req);
+}
+
+void RouteSelector::on_clear_route_mrm(
+  ClearRoute::Request::SharedPtr req, ClearRoute::Response::SharedPtr res)
+{
+  mrm_operating_ = false;
+  mrm_.change_state(now(), RouteState::UNSET);
+
+  // Resume main route using the saved request.
+  if (std::holds_alternative<std::monostate>(main_request_)) {
+    res->status = service_utils::sync_call(cli_clear_route_, req);
+    return;
+  }
+  if (auto request = std::get_if<RoutePointRequest>(&main_request_)) {
+    // NOTE: Clear the waypoint to avoid returning. Remove this once resuming is supported.
+    (**request).waypoints.clear();
+    res->status = service_utils::sync_call(cli_set_waypoint_route_, *request);
+    return;
+  }
+  if (auto request = std::get_if<RouteRequest>(&main_request_)) {
+    res->status = service_utils::sync_call(cli_set_lanelet_route_, *request);
+    return;
+  }
+  RCLCPP_ERROR_STREAM(get_logger(), "unknown main route request");
+}
+
+void RouteSelector::on_set_waypoint_route_mrm(
+  SetWaypointRoute::Request::SharedPtr req, SetWaypointRoute::Response::SharedPtr res)
+{
+  mrm_operating_ = true;
+  res->status = service_utils::sync_call(cli_set_waypoint_route_, req);
+}
+
+void RouteSelector::on_set_lanelet_route_mrm(
+  SetLaneletRoute::Request::SharedPtr req, SetLaneletRoute::Response::SharedPtr res)
+{
+  mrm_operating_ = true;
+  res->status = service_utils::sync_call(cli_set_lanelet_route_, req);
+}
```

Comment:
`mrm_operating_` should be set to `true` only after rerouting success.
Otherwise, it appears that state will also be set if mrm reroute fails. (0:59-1:00 in your demo video)

![image](https://github.com/autowarefoundation/autoware.universe/assets/39142679/202b5e4f-0243-435f-8baf-0620c9d1a85b)

![image](https://github.com/autowarefoundation/autoware.universe/assets/39142679/cdf7af62-f3ef-4060-8622-10119fe7fe1d)